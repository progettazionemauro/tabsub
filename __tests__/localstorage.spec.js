import mockLocalStorage from '../__setup__/localStorage';
import tabsub from '../src/index';

let restoreLocalStorage;
let restoreBroadcast;

beforeAll(() => {
  restoreLocalStorage = mockLocalStorage();
  restoreBroadcast = global.BroadcastChannel;
  global.BroadcastChannel = undefined;
});

beforeEach(() => {
  global.localStorage.removeAll();
});

afterAll(() => {
  restoreLocalStorage();
  global.BroadcastChannel = restoreBroadcast;
});

it('used as fallback strategy', () => {
  const p = tabsub('channel');
  expect(p).toMatchSnapshot();
});

it('posts messages to subscribers', () => {
  const subscriberMockCb = jest.fn(msg => msg);
  const p = tabsub('localChannel');
  const s = tabsub('localChannel');
  s.on(subscriberMockCb);

  p.post('A msg from localstorage');
  p.post('Another one');
  p.post('And a third one');
  expect(subscriberMockCb.mock.calls).toMatchSnapshot();
});

it('posts messages with different types', () => {
  const subscriberMockCb = jest.fn(msg => msg);
  const p = tabsub('localChannel2');
  const s = tabsub('localChannel2');
  s.on(subscriberMockCb);

  p.post('A msg');
  p.post({foo: 'bar'});
  p.post({baz: 1});
  p.post(1);
  p.post();
  p.post(['lovely', 'quz', 1, 0.4]);

  expect(subscriberMockCb.mock.calls).toMatchSnapshot();
});

it('stops and starts the messages on demand', () => {
  const subscriberMockCb = jest.fn(msg => msg);
  const p = tabsub('localChannel3');
  const s = tabsub('localChannel3');
  s.on(subscriberMockCb);

  p.post('A msg');
  s.stop();

  p.post('this will get ignored from the subscriber');
  p.post('And this one');
  s.start();
  p.post('Keep them coming');
  expect(subscriberMockCb.mock.calls).toMatchSnapshot();
});

it('closes the channel', () => {
  const channelName = 'localChannel4';
  const p = tabsub(channelName);
  const message = 'A msg that will be saved';
  p.post(message);

  const retrieved = global.localStorage.getItem(channelName);
  const parsed = JSON.parse(retrieved);
  expect(parsed.channelName).toEqual(channelName);
  expect(parsed.msg).toEqual(message);
  p.close();

  const checkOnceMore = global.localStorage.getItem(channelName);
  expect(checkOnceMore).toBeUndefined();
});

it('shows a warning if localstorage is not supported', () => {
  global.localStorage = undefined;
  global.console = {warn: jest.fn()};
  const p = tabsub('oops');
  expect(p).toMatchSnapshot();
  p.post();
  expect(console.warn).toBeCalled();
});
