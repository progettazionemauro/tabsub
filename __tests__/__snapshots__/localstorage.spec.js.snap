// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posts messages to subscribers 1`] = `
Array [
  Array [
    "A msg from localstorage",
  ],
  Array [
    "Another one",
  ],
  Array [
    "And a third one",
  ],
]
`;

exports[`posts messages with different types 1`] = `
Array [
  Array [
    "A msg",
  ],
  Array [
    Object {
      "foo": "bar",
    },
  ],
  Array [
    Object {
      "baz": 1,
    },
  ],
  Array [
    1,
  ],
  Array [
    Object {},
  ],
  Array [
    Array [
      "lovely",
      "quz",
      1,
      0.4,
    ],
  ],
]
`;

exports[`shows a warning if localstorage is not supported 1`] = `
Object {
  "close": [Function],
  "name": "oops",
  "on": [Function],
  "post": [Function],
  "start": [Function],
  "stop": [Function],
}
`;

exports[`stops and starts the messages on demand 1`] = `
Array [
  Array [
    "A msg",
  ],
  Array [
    "Keep them coming",
  ],
]
`;

exports[`used as fallback strategy 1`] = `
Object {
  "close": [Function],
  "name": "channel",
  "on": [Function],
  "post": [Function],
  "start": [Function],
  "stop": [Function],
}
`;
